### PR TITLE
feat: Add transaction flow and details sections

### PR DESCRIPTION
**Pull request by pedronfigueiredo**
## **Description**

Adds two new sections for the wallet initiated ERC20 token transfer redesigned confirmation.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/27654?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/3220

## **Manual testing steps**

1. Go to this page...
2.
3.

### **After**

Screenshot 2024-10-07 at 11 04 58

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/pages/confirmations/components/confirm/info/hooks/use-token-values.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import { TransactionMeta } from '@metamask/transaction-controller';
+import { useMemo } from 'react';
+import { calcTokenAmount } from '../../../../../../../shared/lib/transactions-controller-utils';
+import { toChecksumHexAddress } from '../../../../../../../shared/modules/hexstring-utils';
+import useTokenExchangeRate from '../../../../../../components/app/currency-input/hooks/useTokenExchangeRate';
+import { useFiatFormatter } from '../../../../../../hooks/useFiatFormatter';
+import { useTokenTracker } from '../../../../../../hooks/useTokenTracker';
+import { SelectedToken } from '../shared/selected-token';
+
+export const useTokenValues = (
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/hooks/use-token-values.ts:10]: Is the target of this PR wrong as all these changes were also in the token send PR?

**Review by matthewwalsh0 — CHANGES_REQUESTED** [review]

File: ui/pages/confirmations/components/confirm/info/hooks/use-token-values.ts (hunk 1 above)
Comment (pedronfigueiredo) [reply ui/pages/confirmations/components/confirm/info/hooks/use-token-values.ts:11]: should be up to date now

File: ui/pages/confirmations/components/confirm/info/token-transfer/token-transfer.tsx [hunk 2]
```diff
@@ -1,8 +1,26 @@
...
+    selectConfirmationAdvancedDetailsOpen,
+  );
+
+  return (
+    <>
+      <SendHeading />
+      <TransactionFlowSection />
+      <TokenDetailsSection />
+      <GasFeesSection />
+      {showAdvancedDetails && <AdvancedDetails />}
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/token-transfer/token-transfer.tsx:21]: Minor, I wonder if it's worth encapsulating the selector of the preference within `AdvancedDetails` itself, we could add a property like `ignorePreference` if we wanted it always present in specific screens.

File: ui/pages/confirmations/components/confirm/info/token-transfer/transaction-flow-section.tsx
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/token-transfer/transaction-flow-section.tsx:1]: Does this component need unit tests?

File: ui/pages/confirmations/components/confirm/info/token-transfer/transaction-flow-section.tsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+export const TransactionFlowSection = () => {
+  const { currentConfirmation: transactionMeta } =
+    useConfirmContext<TransactionMeta>();
+
+  const { value, pending } = useDecodedTransactionData();
+
+  const recipientAddress =
+    value?.data[0].params.find((param) => param.type === 'address')?.value ||
+    '0x0000000000000000000000000000000000000000';
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/token-transfer/transaction-flow-section.tsx:30]: Is this a safe fallback or should we let an error happen?

File: ui/pages/confirmations/components/confirm/info/token-transfer/transaction-flow-section.tsx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import { useConfirmContext } from '../../../../context/confirm';
+import { useDecodedTransactionData } from '../hooks/useDecodedTransactionData';
+import { ConfirmLoader } from '../shared/confirm-loader/confirm-loader';
+import { NameOrAddressDisplay } from './name-or-address-display';
+
+export const TransactionFlowSection = () => {
+  const { currentConfirmation: transactionMeta } =
+    useConfirmContext<TransactionMeta>();
+
+  const { value, pending } = useDecodedTransactionData();
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/token-transfer/transaction-flow-section.tsx:26]: The downside of this is every token transfer will now trigger 4byte and Sourcify requests which previously was only done when the advanced toggle was pressed.

File: ui/pages/confirmations/components/confirm/info/token-transfer/token-details-section.tsx [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  getNetworkConfigurationsByChainId,
+} from '../../../../../../selectors';
+import { useConfirmContext } from '../../../../context/confirm';
+
+export const TokenDetailsSection = () => {
+  const t = useI18nContext();
+  const { currentConfirmation: transactionMeta } =
+    useConfirmContext<TransactionMeta>();
+
+  const chainId = useSelector(getCurrentChainId);
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/token-transfer/token-details-section.tsx:37]: We're actively removing these references now, so can we instead read this from the `currentConfirmation`?

File: ui/pages/confirmations/components/confirm/info/token-transfer/name-or-address-display.tsx [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+import { getPetnamesEnabled } from '../../../../../../selectors';
+
+export const NameOrAddressDisplay = ({ address }: { address: string }) => {
+  const isPetNamesEnabled = useSelector(getPetnamesEnabled);
+  const { displayName, hexAddress } = useFallbackDisplayName(address);
+  const [isNicknamePopoverShown, setIsNicknamePopoverShown] = useState(false);
+  const handleDisplayNameClick = () => setIsNicknamePopoverShown(true);
+  const onCloseHandler = () => setIsNicknamePopoverShown(false);
+
+  if (isPetNamesEnabled) {
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/components/confirm/info/token-transfer/name-or-address-display.tsx:29]: I recall we had plans to remove the toggle for petnames so this won't be needed soon, is it simpler to force this now and remove the alternate code path and maybe even this entire component?

**Review by matthewwalsh0 — CHANGES_REQUESTED** [review]

File: ui/pages/confirmations/components/confirm/info/token-transfer/transaction-flow-section.tsx (hunk 3 above)
Comment (pedronfigueiredo) [reply ui/pages/confirmations/components/confirm/info/token-transfer/transaction-flow-section.tsx:30]: That's a good point, I removed the fallback.

File: ui/pages/confirmations/components/confirm/info/token-transfer/transaction-flow-section.tsx (hunk 4 above)
Comment (pedronfigueiredo) [reply ui/pages/confirmations/components/confirm/info/token-transfer/transaction-flow-section.tsx:26]: That's true... This is the only way I can think of to extract the recipient address. 🤔

**Review by matthewwalsh0 — DISMISSED** [review]

**Review by vinistevam — DISMISSED** [review]

**Review by vinistevam — DISMISSED** [review]

**Review by vinistevam — APPROVED** [review]

**Review by matthewwalsh0 — APPROVED** [review]

**Pull request merged**